### ISSUE TITLE
isGlob process substitution

### ISSUE THREAD

**Issue by thoughtsunificator** [11ty/eleventy#3574]
https://github.com/11ty/eleventy/blob/43919d9291b778b504f065863870049d10aff536/src/Util/ProjectDirectories.js#L146

`npx  @11ty/eleventy  --input=<(echo "foo()")`

It is expected that most CLI expecting an input file also support process substitution.

For example, the following works perfectly as is: `node -p 'fs.readFileSync(process.argv[1])' <(echo foo)`

**Comment by thoughtsunificator** [comment]
I should also note that [the package](https://github.com/micromatch/is-glob) is question is not actively maintained.

**Comment by Ryuno-Ki** [comment]
> I should also note that [the package](https://github.com/micromatch/is-glob) is question is not actively maintained.

I am noticing several of those (but this is an unrelated issue). Given that Eleventy tends to require fewer and fewer packages over the versions I'm positive we can find a solution to this.

**Comment by Ryuno-Ki** [comment]
On my Debian Linux, when I log `process.argv` before it is handed to minimist in

https://github.com/11ty/eleventy/blob/v3.0.0/cmd.cjs#L33

I can see that `<(echo "foo()")` is evaluated to `/dev/fd/63`, that is, a file descriptor.

So I created a small test file

```js
var argv = require("minimist")(process.argv.slice(2));
console.log(argv);
```

in the spirit of the [official example](https://www.npmjs.com/package/minimist#example)

and run it with `node test.cjs --input=<(echo "foo()")` to observe the same result.

I don't see anything specific to handle file descriptors in Minimist's codebase.

**Comment by Ryuno-Ki** [comment]
The Programmatic API shows the same behaviour:

```js
import Eleventy from "@11ty/eleventy"

let elev = new Eleventy("/dev/fd/63");
await elev.write();
```

invoked with `node test.mjs <(echo "foo()")`.

That's going to be a nice rabbit hole to go down to! (After lunch)

**Comment by Ryuno-Ki** [comment]
TL;DR: [fast-glob](https://www.npmjs.com/package/fast-glob) cannot see the file descriptor.

```js
import fg from "fast-glob";

const entries = await fg(["/dev/fd/63"]);
console.log(entries);
```

Long version

1. When constructing, Eleventy is holding the input path in `rawInput`:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L104

2. It is then calling its `init` method:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L104

3. This is invoking its config initialization (among other things)

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L483

4. Within here, the `directories` getter is called:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L234

5. This is setting the `rawInput` to a new `ProjectDirectories` instance:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L301

6. Internally this is deferring to a method:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Util/ProjectDirectories.js#L174

7. On Linux machines, everything is a file (but a file descriptor is not a directory) as asserted by

https://github.com/11ty/eleventy/blob/v3.0.0/src/Util/ProjectDirectories.js#L137

8. Now that the config is initialised it also initialises a writer. `TemplateWriter`, since we're outputting HTML:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L530

9. That one is holding a reference to its Eleventy Config:

https://github.com/11ty/eleventy/blob/v3.0.0/src/TemplateWriter.js#L31

Now we're done with the initialisation side.

Let's look at the write.

1. Our Eleventy instance is invoking the method:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L1256

2. This is deferring the call to the TemplateWriter:

https://github.com/11ty/eleventy/blob/v3.0.0/src/Eleventy.js#L1334

3. Which is collecting the paths:

https://github.com/11ty/eleventy/blob/v3.0.0/src/TemplateWriter.js#L428

4. Which defers to EleventyFiles:

https://github.com/11ty/eleventy/blob/v3.0.0/src/TemplateWriter.js#L140

5. Which is running a glob search:

https://github.com/11ty/eleventy/blob/v3.0.0/src/EleventyFiles.js#L400

6. Which is first querying the globs to search for (our fd string)

https://github.com/11ty/eleventy/blob/v3.0.0/src/EleventyFiles.js#L350

7. Before running a filesystem search using fast-glob:

https://github.com/11ty/eleventy/blob/v3.0.0/src/FileSystemSearch.js#L55

And here we receive an empty array as a result which leads to 0 files written in the end.

**Comment by thoughtsunificator** [comment]
Yes, exactly, and it looks like the glob I mentioned is not part of the issue, indeed.

When passed a file descriptor through process substitution, we end up with the following Template Writer object:
`{ passthroughCopy: Array(0), templates: Array(0) }`
With README.md:
`{ passthroughCopy: Array(0), templates: Array(1) }`

**Comment by thoughtsunificator** [comment]
To help visualize, here's the call tree:

![image](https://github.com/user-attachments/assets/66647c93-da86-4d8d-87d2-b73dc63b418f)

**Comment by thoughtsunificator** [comment]
I believe it's because https://github.com/mrmlnc/fast-glob/blob/7f28b159a4e20b1da6e613beec363f8d5a52cb5a/src/providers/filters/entry.ts#L52 checks if it's a file and Node.js fs API ([isFile](https://nodejs.org/api/fs.html#direntisfile)) returns false although it will return true when [isFIFO](https://nodejs.org/api/fs.html#direntisfifo) is called.

**Comment by zachleat** [comment]
Related https://github.com/11ty/eleventy/pull/3680
